### DISCUSSION TITLE
Valtio promises not resolving inside r3f's Canvas when using next.js

### DISCUSSION THREAD

**Discussion by sheunglaili** [pmndrs/valtio#917] [Bug report]
### Bug Description

`TypeError: dispatcher.use is not a function` thrown when trying to access snapshot using `useSnapshot` with state containing a promise field in a child component to r3f's Canvas. 

I have try reproducing in plain react but do not encounter this issue.

### Reproduction Link

https://codesandbox.io/p/devbox/r3f-valtio-promise-issue-nextjs-p8c533?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clxwclyyj0006356ja6919jy9%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clxwclyyj0002356jzvzw1iey%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clxwclyyj0004356jksv9d72c%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clxwclyyj0005356jfp1q247j%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clxwclyyj0002356jzvzw1iey%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxwclyyi0001356j4pem7fgp%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clxwclyyj0002356jzvzw1iey%2522%252C%2522activeTabId%2522%253A%2522clxwclyyi0001356j4pem7fgp%2522%257D%252C%2522clxwclyyj0005356jfp1q247j%2522%253A%257B%2522id%2522%253A%2522clxwclyyj0005356jfp1q247j%2522%252C%2522activeTabId%2522%253A%2522clxwcqfms00gi356jfldx8ro8%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clxwcqfms00gi356jfldx8ro8%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clxwclyyj0004356jksv9d72c%2522%253A%257B%2522id%2522%253A%2522clxwclyyj0004356jksv9d72c%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clxwclyyj0003356jmu00wfzq%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clxwclyyj0003356jmu00wfzq%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D

**Comment by dai-shi** [comment]
Does r3f support React 19 yet?

**Reply by sheunglaili** [reply]
as far as I can see, it supports React 18 but not yet with React 19 https://github.com/pmndrs/react-three-fiber/blob/master/packages/fiber/CHANGELOG.md#800

**Reply by dai-shi** [reply]
That explains. `use` isn't in React 18.

**Reply by sheunglaili** [reply]
correct me if I am wrong but I think the sandbox is using valtio version 1.13.2 which didn't use `use` ?

**Reply by sheunglaili** [reply]
sorry just found it: https://github.com/pmndrs/valtio/blob/9593184ad94928e54382bfd0e0beb7d1416eb6fe/src/react.ts#L24

**Reply by sheunglaili** [reply]
so if I were to use valtio promise, I would have to either downgrade to 1.7 or use 2.0 ?

**Reply by dai-shi** [reply]
In valtio v1, if `use` doesn't exist, it will use a fallback version.
https://github.com/pmndrs/valtio/blob/9593184ad94928e54382bfd0e0beb7d1416eb6fe/src/vanilla.ts#L88-L103

> so if I were to use valtio promise, I would have to either downgrade to 1.7 or use 2.0 ?

If you need to use valtio promise with the current r3f (React 18), you have to use Valtio v1. v1.13.2 should work.

**Reply by sheunglaili** [reply]
hmmm, I am currently using v1.13.2 but still get the `TypeError: dispatcher.use is not a function` when using valtio promises as seen in the codesandbox🤔

**Reply by dai-shi** [reply]
You are probably using React 19 or something. Make sure to use React 18.

**Reply by dai-shi** [reply]
Or, if you use Valtio v2, you shouldn't see `TypeError: dispatcher.use is not a function` error at least.

**Reply by sheunglaili** [reply]
![Screenshot 2024-06-28 at 16 11 45](https://github.com/pmndrs/valtio/assets/31618202/44026b87-1421-4775-b5fe-843cf1810ec3)

if you look at the codesandbox, it's still running on react 18

**Reply by dai-shi** [reply]
Ah, I think next ignores what you have in package.json, and they just use their "bundled" React. Can you try it without Next.js?
